source/fem-dem/gls_sharp_navier_stokes.cc

```diff
@@ -2516,431 +2642,414 @@ GLSSharpNavierStokesSolver<dim>::sharp_edge()
-                          1e-16 * dr)
+
+                      if (component_i == dim &&
+                          this->locally_owned_dofs.is_element(
+                            global_index_overwrite))
                         {
-                          // If the matrix entry on the diagonal of this DOF is
-                          // close to zero, check if all the cells close are
-                          // cut. If it's the case, the DOF is a dummy DOF.
-                          active_neighbors_set =
-                            LetheGridTools::find_cells_around_cell<dim>(
-                              vertices_to_cell, cell);
-                          for (unsigned int m = 0;
-                               m < active_neighbors_set.size();
-                               m++)
+                          // Applied equation on dof that have no equation
+                          // defined for them. those DOF become Dummy dof. This
+                          // is usefull for high order cells or when a dof is
+                          // only element of cells that are cut.
+                          unsigned int global_index_overwrite =
+                            local_dof_indices[i];
+                          bool dummy_dof = true;
+
+                          // To check if the pressure dof is a dummy. first
+                          // check if the matrix entry is close to 0.
+                          if (abs(this->system_matrix.el(
+                                global_index_overwrite,
+                                global_index_overwrite)) <= 1e-16 * dr)
                             {
-                              const auto &cell_3 = active_neighbors_set[m];
-                              cell_3->get_dof_indices(local_dof_indices_3);
-                              for (unsigned int o = 0;
-                                   o < local_dof_indices_3.size();
-                                   ++o)
+                              // If the matrix entry on the diagonal of this DOF
+                              // is close to zero, check if all the cells close
+                              // are cut. If it's the case, the DOF is a dummy
+                              // DOF.
+                              active_neighbors_set =
+                                LetheGridTools::find_cells_around_cell<dim>(
+                                  vertices_to_cell, cell);
+                              for (unsigned int m = 0;
+                                   m < active_neighbors_set.size();
+                                   m++)
                                 {
-                                  if (global_index_overwrite ==
-                                      local_dof_indices_3[o])
+                                  const auto &cell_3 = active_neighbors_set[m];
+                                  cell_3->get_dof_indices(local_dof_indices_3);
```

Comment:
I would rename cell_3 to something that is more verbose. "neighbhor_cell" or something like this.